### PR TITLE
Refactor SZA and cos(SZA) generation to reduce duplicate computations

### PR DESCRIPTION
**Pull request by djhoese**
This is another step towards better performance regarding #1902. I haven't figured everything out yet, but this PR does seem to help a little bit. There is a small amount of duplicate logic between satpy and pyorbital regarding the `sun_zenith_angle` function which was just `np.rad2deg(np.arccos(cos_zen))`. As pointed out in #1902, I noticed that cos(SZA) calculations for the sunz corrector modifier were not being shared with the rayleigh correction which also needs to calculate SZA. This PR refactors the relevant functions and wraps parts in map_blocks to allow dask to reuse calculations when possible.

 - [ ] Closes #xxxx 
 - [ ] Tests added

### CODE REVIEW

**Comment by djhoese** [comment]
I should note that this PR reduced the overall memory usage of the AHI and ABI `true_color` generation. For ABI is went from ~20GB to ~16GB. Note this was with my special test script which does not use a writer (no geotiff saving).

File: satpy/modifiers/angles.py [hunk 1]
```diff
@@ -332,3 +367,49 @@ def _get_sensor_angles_wrapper(lons, lats, start_time, sat_lon, sat_lat, sat_alt
...
+        grad_factor = grad_factor.clip(0.)
+    else:
+        # Use constant value (the limit) for larger zenith angles
+        grad_factor = 1.
+    # corr[cos_zen <= limit_cos] = grad_factor / limit_cos
+    # corr = corr.where(cos_zen > limit_cos, grad_factor / limit_cos)
+    corr = np.where(cos_zen > limit_cos, corr, grad_factor / limit_cos)
+    # Force "night" pixels to 0 (where SZA is invalid)
+    corr[np.isnan(cos_zen)] = 0
+    # corr = corr.where(cos_zen.notnull(), 0)
```
Comment (mraspaud) [inline satpy/modifiers/angles.py:413]: Commented code

**Comment by djhoese** [comment]
"Why leave the commented code?"

Because I made this PR after 10 hours of working on this as a "hack" that then turned into an actual PR. I'll clean it up today.

**Comment by mraspaud** [comment]
Thanks

**Comment by djhoese** [comment]
> Also, it feels like most of the wrapping could be done as a decorator, an be done in pyorbital.

@mraspaud You mean do `map_blocks` as a decorator? Does that work?

**Comment by djhoese** [comment]
Oh you mean make a decorator in pyorbital that handles dask inputs with map_blocks or if passed numpy arrays then calls it directly. Right?

**Comment by djhoese** [comment]
@mraspaud I think I've addressed all your concerns except for moving some of this dask support to pyorbital. Pyorbital doesn't currently have any dask or xarray support as far as I know.

**Comment by mraspaud** [comment]
> Oh you mean make a decorator in pyorbital that handles dask inputs with map_blocks or if passed numpy arrays then calls it directly. Right?

yes

**Comment by djhoese** [comment]
Obviously that has to be done in a separate PR (on pyorbital) but would you be OK with this being merged as-is and then working on pyorbital later?

**Comment by mraspaud** [comment]
> Obviously that has to be done in a separate PR (on pyorbital) but would you be OK with this being merged as-is and then working on pyorbital later?

That's ok, let's hope we have the time to fix pyorbital soon. Do we have an issue for that? otherwise it might be good to create one now, so that we don't forget this too easily

File: satpy/modifiers/angles.py [hunk 2]
```diff
@@ -274,8 +275,20 @@ def get_satellite_zenith_angle(data_arr: xr.DataArray) -> xr.DataArray:
     return satz
 
 
+def get_cos_sza(data_arr: xr.DataArray) -> xr.DataArray:
+    """Generate the cosine of the solar zenith angle for the provided data.
+
+    Returns:
+        DataArray with the same shape as ``data_arr``. Units are radians.
```
Comment (pnuu) [inline satpy/modifiers/angles.py:282]: Cosine of something is not in any angle units. It is the ratio of the length of cathet opposite to the angle and the length of the hypotenuse.

**Review by pnuu — APPROVED** [review]
LGTM. Just one clarification to docstring.

File: satpy/modifiers/angles.py (hunk 2 above)
Comment (djhoese) [reply satpy/modifiers/angles.py]: :man_facepalming: Done. Thanks.

**Comment by djhoese** [comment]
@mraspaud An issue has been created: https://github.com/pytroll/pyorbital/issues/88

**Review by mraspaud — APPROVED** [review]
LGTM

**Pull request merged**